### ISSUE TITLE
Add the ability to specify additional edn readers for config files?

### ISSUE THREAD

**Issue by jskrzypek** [thheller/shadow-cljs#627]
Would it be possible to pass in extra readers to the `edn/read-string :readers` map? I'm imagining something could be pulled in via the `:deps :aliases` (but relatively new to clojure/clojurescript, so maybe that's not feasible)

My problem is that we have several cljs apps that are running in production on different nodes in our cluster and I want to dockerize shadow-cljs processes for our local dev versions. Currently most of us are just running shadow-cljs from our shells or emacs clients, and I want to be able to have docker-compose specify via environment variables which ports shadow-cljs should host the app on as well as opening the nrepl on, but also be able to provide a default if the process is run locally without forcing people to specify the port in the local env, as the example below would require.

I thought that maybe functionality could be added so I could pass in a map of `:readers` in the `:deps :aliases` like I can with `:nrepl`, where I could have a library that provides a reader function, and then I could specify something like `#custom/env-with-default "HTTP_PORT:8231"` that would check System.getenv for `$HTTP_PORT` like `#shadow/env` but fall back to `8231` if no env var with that name is defined, something like

```clj
(defn getenv-with-default 
  [envname-with-default]
  (let [[envname default] (clojure.string/split envname-with-default #":" 1)]
     (or (System/getenv envname) default)))
```

Would something like that be possible?

-----------------------------

### Example of the currently available solution
`/app-client-1/shadow-cljs.edn`
```edn
{:deps {:aliases [:nrepl]}
 :source-paths ["src" "test"]
 :nrepl  {:port #shadow/env "NREPL_PORT_1")
 :builds {:app {:devtools {:http-port #shadow/env "HTTP_PORT_1"}}}}
```

`/app-client-2/shadow-cljs.edn`
```edn
{:deps {:aliases [:nrepl]}
 :source-paths ["src" "test"]
 :nrepl  {:port #shadow/env "NREPL_PORT_2")
 :builds {:app {:devtools {:http-port #shadow/env "HTTP_PORT_2"}}}}
```

`/docker-compose.yml`
```yml
services:
  app-client-1:
    restart: always
    ports:
      - 127.0.0.1:8241:8230
      - 127.0.0.1:8291:8280
    environment:
      NREPL_PORT_1: "8241"
      HTTP_PORT_1: "8291"
    build:
      context: '.'
      dockerfile: 'app-client-1/Dockerfile.local'
    volumes:
      - './shared-components:/app/shared-components'
      - './app-client-1/package.json:/app/app-client-1/package.json'
      - './app-client-1/shadow-cljs.edn:/app/app-client-1/shadow-cljs.edn'
      - './app-client-1/public/assets:/app/app-client-1/public/assets'
      - './app-client-1/public/index.html:/app/app-client-1/public/index.html'
      - './app-client-1/src:/app/app-client-1/src'
    command:
      [
        "npx",
        "shadow-cljs",
        "watch",
        "app",
      ]
  app-client-2:
    restart: always
    ports:
      - 127.0.0.1:8242:8231
      - 127.0.0.1:8292:8281
    environment:
      NREPL_PORT_2: "8242"
      HTTP_PORT_2: "8292"
    build:
      context: '.'
      dockerfile: 'app-client-2/Dockerfile.local'
    volumes:
      - './shared-components:/app/shared-components'
      - './app-client-2/package.json:/app/app-client-2/package.json'
      - './app-client-1/shadow-cljs.edn:/app/app-client-2/shadow-cljs.edn'
      - './app-client-1/public/assets:/app/app-client-2/public/assets'
      - './app-client-1/public/index.html:/app/app-client-2/public/index.html'
      - './app-client-1/src:/app/app-client-2/src'
    command:
      [
        "npx",
        "shadow-cljs",
        "watch",
        "app",
      ]
```

**Comment by thheller** [comment]
No, that is currently not supported. I won't be adding support for custom readers but I might be open to adding a `#shadow/env ["KEY" "default-value"]` variant.

However ... setting the `:nrepl` port in the first place is already something that should never be done.

`shadow-cljs` writes the port it uses to `.shadow-cljs/nrepl.port` on startup and editors are supposed to read that file when they need the port. Unfortunately most tools either don't support that method at all or only support the hardcoded `.nrepl-port` file from leiningen. I wish tools would just allow configuring this more ...

**Comment by thheller** [comment]
Curious why these are 2 separate projects? They seem to share a lot? Why not one project with 2 builds?

**Comment by jskrzypek** [comment]
Thanks for the quick reply. 

> Curious why these are 2 separate projects? They seem to share a lot? Why not one project with 2 builds?

Ultimately these clients get served by distinct kubernetes pods in our GKE clusters, and they need to know how to route to one another, as well as to our backend APIs which are running in other pods in the same clusters. Also not every cluster we spin up will actually have every client running in it.  

I suppose we could have one omnibus shadow-cljs.edn that knows about every client in our monorepo and is only used for local dev flows, but the rest of our stack for local dev is manged by our docker-compose.yml, and I worry that doing something different will lead to unexpected gotchas down the road when it comes to CI/CD phases of our dev cycle. This whole issue is actually resulting from the inverse of that problem: someone modified something and only fixed the production version, which broke other peoples' local workflows. 

> No, that is currently not supported. I won't be adding support for custom readers but I might be open to adding a #shadow/env ["KEY" "default-value"] variant.

The default option would definitely be a nice add! :pray: 

-----------

Also, I just actually implemented & tested what I wrote up as the current solution in our real-world project and realized that what I was expecting to work doesn't because Java's `System.getenv()` is returning a string, so I get the following spec error. Do you have any suggestions how to overcome this?

```
Invalid configuration
-- Spec failed --------------------
  {:output-dir ...,
   :compiler-options ...,
   :build-hooks ...,
   :modules ...,
   :devtools
   {:http-root ...,
    :http-port "8280",
               ^^^^^^
    :after-load ...,
    :preloads ...},
   :js-options ...,
   :build-id ...,
   :asset-path ...,
   :target ...}
should satisfy
  pos-int?```

**Comment by thheller** [comment]
Wouldn't it be easier to "generate" the shadow-cljs.edn config as part of the docker setup? I really want the config file to stay pure data and putting too much logic in them defeats that purpose.

Since they are pure data you can `read-string` them, then `update-in` or `assoc-in` whatever you need to set and then `(spit (pr-str updated-config) ...)` before actually starting shadow-cljs. Since you are using `deps.edn` anyways that would be a couple lines of code allowing for any customization you might need?

**Comment by jskrzypek** [comment]
Oh, that's an interesting idea! I'll try that.

Your point about pure data is a good one, but since shadow already provides
`#shadow/env`, it feels like an awkward/arbitrary limitation to not be able
to pull in a number from the env.

Perhaps adding an explicit alternative like `#shadow/env-number` would help
keep the edn clean and declarative?

**Comment by jskrzypek** [comment]
If you like the idea, I'd be happy to make a PR for it (since it'd be very straightforward)

**Comment by thheller** [comment]
I let someone talk me into merging `#shadow/env`. I regret that decision. `--config-merge` I added since  is a much better way to take care of "customizing" configs but that only allows merging build config options. Maybe there should be a variant for overall config. There is also [~/.shadow-cljs/config.edn](https://shadow-cljs.github.io/docs/UsersGuide.html#user-config)

`#shadow/env-as-int` `#shadow/env-as-bool` etc are not something I'm a big fan of. It just never ends. At some point it makes more sense to use "code" to generate the config instead. I do not want the config to become code. `#shadow/env` is already a problem if tools like Cursive want to start reading the config.

**Comment by jskrzypek** [comment]
You're right, I hadn't realized other tools were depending on the config
outside of shadow...

Now I'm actually inclined to just use `--config-merge` in my docker build
and just create an edn file that will override what I need

**Comment by thheller** [comment]
`--config-merge` doesn't work for the settings you used in your example though. It only works for actual "builld" config and the `:devtools :http-root` is no longer build specific and deprecated (now `:dev-http`).

**Comment by tlrobinson** [comment]
I'm looking for a way to change `:dev-http`'s `:proxy-url` without having to generate a config file (e.x. for a Docker container where the hostname of the proxy target may change)

AFAICT `#shadow/env` is the only way to do that. The only downside for me is the lack of a way to specify a default.

If you're looking for prior art, `#duct/env` would be nice to match: https://cljdoc.org/d/duct/core/0.8.0/api/duct.core.env

**Comment by thheller** [comment]
Fixed in `2.10.4`.

I added some [examples](https://shadow-cljs.github.io/docs/UsersGuide.html#shadow-env) to the Guide.
